### DISCUSSION TITLE
Can Quarto generate HTML output as a SINGLE file (and not an HTML file + folder)

### DISCUSSION THREAD

**Discussion by jhchou** [quarto-dev/quarto-cli#2666] [Q&A] [answered]
I use `flexdashboard` extensively to report to my non-technical colleagues, and one of the great things is that I can distribute a single HTML file (with all the graphs/plots internal to the HTML file) by email and just ask them to save the file attachment and open it with a web browser. (And even then, they sometimes have trouble...)

For Quarto HTML output, it seems to save as an HTML file and folders containing the figures and libraries (CSS, JS) to display. It's going to be much harder to distribute the reports to my non-technical colleagues if I have to ask them to save BOTH files to a single location, uncompressing the .gz folder of files, etc.

I much prefer the potential interactivity of HTML output, compared with, say, PDF or MS Word/docx.

Is there a way to save the output in one giant HTML file?

**Comment by dragonstyle** [comment, marked as answer]
If you set `self-contained: true` in the document front matter, Quarto should place all the references within the HTML file itself (it sounds like what you're looking for. More here:

https://quarto.org/docs/output-formats/html-basics.html#self-contained

**Reply by jhchou** [reply]
Perfect, thank you very much!

(Time to re-read all the documentation I guess)

**Reply by mcanouil** [reply]
@dragonstyle should you not be highlighting the `embed-resources` option? 😏
From <https://quarto.org/docs/reference/formats/html.html>:
> `self-contained`: Produce a standalone HTML file with no external dependencies. Note that this option has been deprecated in favor of `embed-resources`.

Maybe worth updating the different examples of the documentation to be consistent with that deprecation notice (which is a consequence of Pandoc deprecation notice on the same argument).

**Reply by dragonstyle** [reply]
Correct!

**Reply by cderv** [reply]
I believe we need to advice to use `embed-resources` now and no more `self-contained`. We should update our doc.
@dragonstyle I have opened https://github.com/quarto-dev/quarto-web/issues/422 to track

**Reply by dmenne** [reply]
`embed-resources` works, but it creates separate html-files  for each chapter of the book, and so it is correctly no longer named `self-contained`. Distributing this by email to reviewers is not acceptable, because most of them get confused by multiple files.  Don't ask to put is on a website, external customers get crazy when you put their holy cows (or patient data) on a website.

In knitr/rmarkdown, `bookdown::html_document2` created a single html-document2 for multiple chapters, which I find very helpful for distribution. The only alternative now is epub. It would be perfect when ebub reader with commenting functions were installed on corporate computers, but sadly this is not the case.

MS-Word might work and even be more acceptable as output with review-capabilities, but it has other limitations, i.e. needs manual editing to avoid "Why is this table to ugly"

So what is the solution when you have a large document you want to split up for editing, and have single (really!).

(My dream: epub with Word-like annotation features, resulting in ONE file)

**Reply by cderv** [reply]
@dmenne Thanks for your comment. I believe your use case is a bit different than the one here, and would be worth discussing in its own thread. Anyway, I'll try to give some insights. 

`bookdown::html_document2()` is not a book format in **bookdown**, it is designed as a single document format. So indeed different than Quarto book. **bookdown** uses a workflow that has its limitation but allow this : it merges every source in one document before rendering to output format with Pandoc. Quarto does not do that - This is discussed in https://github.com/quarto-dev/quarto-cli/discussions/3577 and This is discussed in https://github.com/quarto-dev/quarto-cli/discussions/1045 and This is discussed in https://github.com/quarto-dev/quarto-cli/discussions/431

To have a single document before rendering, you could maybe leverage [include](https://quarto.org/docs/authoring/includes.html#overview) shortcode to create one single document before rendering. Then you would render using the html format. Same idea of splitting a file in several using **knitr** with child document if you know it. 

I find the idea interesting to able to this inclusion at the project level though... 

Anyway, single file output for book project are currently not HTML but PDF or Word.

**Reply by cscheid** [reply]
> `embed-resources` works, but it creates separate html-files for each chapter of the book, and so it is correctly no longer named `self-contained`.

It's likely that our documentation is out of date in spots since this was a recent Pandoc change, I apologize in advance. But `embed-resources` is not a substitute for `self-contained`; what happened was that `self-contained` was split into two options: `embed-resources` and `standalone`. So could it be that adding `standalone: true` solves the problem for you?

**Reply by cderv** [reply]
@cscheid just to clarify for me: Are we supporting creating one single HTML file from a Quarto book ? 

I did not mention `standalone: true` because I did not see it as related with this usage - My understanding is than `standalone: true` is the default in all our format. So I assume it was set already. 

To clarify about `bookdown::html_document2()`  mentioned above, this is a single file format. Equivalent to our `format: html`, but with **bookdown** project, you are able to write in several Rmd files and it will produce a single HTML file at then with this format. `bookdown::gitbook()` or other book format will produce several html files like in Quarto book.

**Reply by cscheid** [reply]
@cderv No - I am the one confused, and had missed the fact that this is a book we're taking about. I apologize.
